apps/emqx_node_rebalance/src/emqx_node_rebalance_purge.erl

```diff
@@ -0,0 +1,234 @@
+    code_change/4
+]).
+
+-export_type([
+    start_opts/0,
+    start_error/0,
+    stop_error/0
+]).
+
+%%--------------------------------------------------------------------
+%% APIs
+%%--------------------------------------------------------------------
+
+-define(DEFAULT_PURGE_RATE, 500).
+-define(ENABLE_KIND, purge).
+
+%% gen_statem states
+-define(disabled, disabled).
+-define(purging, purging).
+-define(cleaning_data, cleaning_data).
+
+-type start_opts() :: #{
+    purge_rate => pos_integer()
+}.
+-type start_error() :: already_started.
+-type stop_error() :: not_started.
+-type stats() :: #{
+    initial_sessions := non_neg_integer(),
+    current_sessions := non_neg_integer(),
+    purge_rate := pos_integer()
+}.
+-type status() :: {enabled, stats()} | disabled.
+
+-spec start(start_opts()) -> ok_or_error(start_error()).
+start(StartOpts) ->
+    Opts = maps:merge(default_opts(), StartOpts),
+    gen_statem:call(?MODULE, {start, Opts}).
+
+-spec stop() -> ok_or_error(not_started).
+stop() ->
+    gen_statem:call(?MODULE, stop).
+
+-spec status() -> status().
+status() ->
+    gen_statem:call(?MODULE, status).
+
+-spec start_link() -> startlink_ret().
+start_link() ->
+    gen_statem:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+%%--------------------------------------------------------------------
+%% gen_statem callbacks
+%%--------------------------------------------------------------------
+
+callback_mode() -> handle_event_function.
+
+%% states: disabled, purging, cleaning_data
+
+init([]) ->
+    {ok, disabled, #{}}.
+
+%% start
+handle_event({call, From}, {start, Opts}, ?disabled, #{} = Data) ->
+    ok = enable_purge(),
+    ?SLOG(warning, #{
+        msg => "cluster_purge_started",
+        opts => Opts
+    }),
+    NewData = init_data(Data, Opts),
+    {next_state, ?purging, NewData, [
+        {state_timeout, 0, purge},
+        {reply, From, ok}
+    ]};
+handle_event({call, From}, {start, _Opts}, _State, #{}) ->
+    {keep_state_and_data, [{reply, From, {error, already_started}}]};
+%% stop
+handle_event({call, From}, stop, ?disabled, #{}) ->
+    {keep_state_and_data, [{reply, From, {error, not_started}}]};
+handle_event({call, From}, stop, _State, Data) ->
+    ok = disable_purge(),
+    ?SLOG(warning, #{msg => "cluster_purge_stopped"}),
+    {next_state, disabled, deinit(Data), [{reply, From, ok}]};
+%% status
+handle_event({call, From}, status, ?disabled, #{}) ->
+    {keep_state_and_data, [{reply, From, disabled}]};
+handle_event({call, From}, status, State, Data) ->
+    Stats = maps:with(
+        [
+            initial_sessions,
+            current_sessions,
+            purge_rate
+        ],
+        Data
+    ),
+    {keep_state_and_data, [
+        {reply, From, {enabled, Stats#{state => State}}}
+    ]};
+%% session purge
+handle_event(
+    state_timeout,
+    purge,
+    ?purging,
+    #{
+        purge_rate := PurgeRate
+    } = Data
+) ->
+    case emqx_eviction_agent:status() of
```

Comment:
This reports local counts, status, multicall is needed?

Reply:
Excellent catch.  The test was passing due to the symmetry in setting up the clients.  :see_no_evil: